tasks/agent.py

```diff
@@ -370,6 +364,28 @@ def hacky_dev_image_build(
                 latest_release = ver
         base_image = f"gcr.io/datadoghq/agent:{latest_release}"
 
+    # Extract the python library of the docker image
+    with tempfile.TemporaryDirectory() as extracted_python_dir:
+        ctx.run(
+            f"docker run --rm '{base_image}' bash -c 'tar --create /opt/datadog-agent/embedded/{{bin,lib,include}}/*python*' | tar --directory '{extracted_python_dir}' --extract"
+        )
+
+        os.environ["DELVE"] = "1"
+        os.environ["LD_LIBRARY_PATH"] = (
+            os.environ.get("LD_LIBRARY_PATH", "") + f":{extracted_python_dir}/opt/datadog-agent/embedded/lib"
+        )
+        build(
+            ctx,
+            cmake_options=f'-DPython3_ROOT_DIR={extracted_python_dir}/opt/datadog-agent/embedded -DPython3_FIND_STRATEGY=LOCATION',
+        )
+        ctx.run(
```

Comment:
Can you explain what this perl command does? Ideally could it be written using python as not all developers are familiar with `perl`.

Reply:
It turns out that the `libdatadog-agent-three.so` library embeds the **absolute** path of the python runtime used at compile time.
This is an issue because it hasn’t the same location at runtime (inside the docker image).

Here is the symptom we see without this `perl` command:
```
2023-09-07 17:47:01 UTC | CORE | INFO | (pkg/collector/python/init.go:321 in resolvePythonExecPath) | Using '' as Python home
2023-09-07 17:47:01 UTC | CORE | WARN | (pkg/collector/python/init.go:346 in resolvePythonExecPath) | Python home is empty. Inferring interpreter path from binary in path.
2023-09-07 17:47:01 UTC | CORE | INFO | (pkg/collector/python/init.go:399 in Initialize) | Initializing rtloader with Python 3 
Python path configuration:
  PYTHONHOME = '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../..'
  PYTHONPATH = (not set)
  program name = '/opt/datadog-agent/embedded/bin/python3'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = '/opt/datadog-agent/embedded/bin/python3'
  sys.base_prefix = '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../..'
  sys.base_exec_prefix = '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../..'
  sys.platlibdir = 'lib'
  sys.executable = '/opt/datadog-agent/embedded/bin/python3'
  sys.prefix = '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../..'
  sys.exec_prefix = '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../..'
  sys.path = [
    '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../../lib/python39.zip',
    '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../../lib/python3.9',
    '/tmp/tmp4_e1e524/opt/datadog-agent/embedded/lib/python3.9/../../lib/python3.9/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007fc985a9b6c0 (most recent call first):
<no Python frame>
```

Rather than patching the binary `libdatadog-agent-three.so` file, I tried to explicitly set `PYTHONHOME` and/or `PYTHONPATH`, but it didn’t work.
So, I resigned myself to patch the binary.

If the “wrong” absolute path prefixed by `/tmp/tmp…` had been in a `RPATH`, I could have used `patchelf`.
But unfortunately, I found this path in the `.rodata` section:

```
$ readelf -p .rodata dev/lib/libdatadog-agent-three.so | grep embedded/lib
  [     8]  /tmp/tmp52tsdpuf/opt/datadog-agent/embedded/lib/python3.9/../..
```

and `patchelf` don’t have any facility to patch this part of an ELF file.

So, the goal of this `perl` command is to remove the `/tmp/tmp…` prefix in front of the `/opt/datadog-agent/embedded/lib/python3.9/../..` string inside the ELF object and to pad it with NUL bytes at the end to ensure that what’s after this string in the ELF object isn’t shifted and everything remains at the same offset.